File: back/src/whombat/routes/clips.py
"""REST API routes for clips."""

from uuid import UUID, uuid4

from fastapi import APIRouter, Depends

from whombat import api, schemas
from whombat.filters.clips import ClipFilter
from whombat.filters.recordings import UUIDFilter as RecordingUUIDFilter
from whombat.routes.dependencies import Session
from whombat.routes.types import Limit, Offset

__all__ = [
    "clips_router",
]


clips_router = APIRouter()


@clips_router.get(
    "/",
    response_model=schemas.Page[schemas.Clip],
)
async def get_clips(
    session: Session,
    limit: Limit = 10,
    offset: Offset = 0,
    filter: ClipFilter = Depends(ClipFilter),  # type: ignore
    sort_by: str = "-created_on",
):
    """Get a page of clips."""
    tasks, total = await api.clips.get_many(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
        sort_by=sort_by,
    )
    return schemas.Page(
        items=tasks,
        total=total,
        limit=limit,
        offset=offset,
    )


@clips_router.post(
    "/",
    response_model=list[schemas.Clip],
)
async def create_clips(
    session: Session,
    data: list[tuple[UUID, schemas.ClipCreate]],
):
    """Create multiple clips."""
    recording_uuids = list(set(recording_uuid for recording_uuid, _ in data))
    recordings, _ = await api.recordings.get_many(
        session,
        filters=[
            RecordingUUIDFilter(
                isin=recording_uuids,
            )
        ],
        limit=None,
        sort_by=None,
    )
    recording_mapping = {recording.uuid: recording for recording in recordings}
    clips = await api.clips.create_many_without_duplicates(
        session,
        [
            dict(
                uuid=uuid4(),
                recording_id=recording_mapping[recording_uuid].id,
                start_time=clip.start_time,
                end_time=clip.end_time,
            )
            for (recording_uuid, clip) in data
        ],
        return_all=True,
    )
    await session.commit()
    return clips


@clips_router.delete(
    "/detail/",
    response_model=schemas.Clip,
)
async def delete_clip(
    session: Session,
    clip_uuid: UUID,
):
    """Delete a clip."""
    clip = await api.clips.get(session, clip_uuid)
    clip = await api.clips.delete(session, clip)
    await session.commit()
    return clip

File: front/src/lib/api/clips.ts
import { AxiosInstance } from "axios";
import { z } from "zod";

import * as schemas from "@/lib/schemas";
import type * as types from "@/lib/types";

import { GetMany, Page } from "./common";

const DEFAULT_ENDPOINTS = {
  createMany: "/api/v1/clips/",
  getMany: "/api/v1/clips/",
  delete: "/api/v1/clips/detail/",
};

export function registerClipAPI(
  api: AxiosInstance,
  endpoints: typeof DEFAULT_ENDPOINTS = DEFAULT_ENDPOINTS,
) {
  async function getMany(
    query: types.GetMany & types.ClipFilter,
  ): Promise<types.Page<types.Clip>> {
    const params = GetMany(schemas.ClipFilterSchema).parse(query);
    const response = await api.get(endpoints.getMany, {
      params: {
        limit: params.limit,
        offset: params.offset,
        sort_by: params.sort_by,
        recording__eq: params.recording?.uuid,
        dataset__eq: params.dataset?.uuid,
        start_time__lt: params.start_time?.lt,
        start_time__gt: params.start_time?.gt,
        end_time__lt: params.end_time?.lt,
        end_time__gt: params.end_time?.gt,
        feature__name: params.feature?.name,
        feature__lt: params.feature?.lt,
        feature__gt: params.feature?.gt,
      },
    });
    return Page(schemas.ClipSchema).parse(response.data);
  }

  async function createMany(data: types.ClipCreateMany): Promise<types.Clip[]> {
    const response = await api.post(endpoints.createMany, data);
    return z.array(schemas.ClipSchema).parse(response.data);
  }

  async function deleteClip(clip: types.Clip): Promise<types.Clip> {
    const response = await api.delete(endpoints.delete, {
      params: { clip_uuid: clip.uuid },
    });
    return schemas.ClipSchema.parse(response.data);
  }

  return {
    getMany,
    createMany,
    delete: deleteClip,
  } as const;
}
